reframe/frontend/executors/policies.py

```diff
@@ -221,337 +235,404 @@ def exit(self):
+                stats[t.policy_stage][t.check.current_partition.fullname].append(t)
 
-    def on_task_success(self, task):
-        msg = f'{task.check.info()} [{task.pipeline_timings_basic()}]'
-        self.printer.status('OK', msg, just='right')
-        getlogger().verbose(f'==> timings: {task.pipeline_timings_all()}')
+        print(f"Tasks in wait: {len(stats['wait'])}")
+        phases = ['ready_to_compile', 'compiling', 'ready_to_run', 'running', 'completed']
+        for phase in phases:
+            print(f"Tasks in {phase}:")
+            for part in stats[phase]:
+                print(f"  {part}: {len(stats[phase][part])}")
 
-        # Update reference count of dependencies
-        for c in task.testcase.deps:
-            # NOTE: Restored dependencies are not in the task_index
-            if c in self._task_index:
-                self._task_index[c].ref_count -= 1
+    def exit(self):
+        self.printer.separator('short single line',
+                               'waiting for spawned checks to finish')
+        while self._current_tasks:
+            # print()
+            # self.print_state_of_tasks(self._current_tasks)
+            try:
+                self._poll_tasks()
+                num_running = sum(
+                    1 if t.policy_stage in ['running', 'compiling']
+                    else 0 for t in self._current_tasks
+                )
+                self.advance_all(self._current_tasks)
+                _cleanup_all(self._retired_tasks, not self.keep_stage_files)
+                if num_running:
+                    self._pollctl.running_tasks(num_running).snooze()
+            except ABORT_REASONS as e:
+                self._failall(e)
+                raise
 
-        self._retired_tasks.append(task)
+        self.printer.separator('short single line',
+                               'all spawned checks have finished\n')
 
-    def on_task_exit(self, task):
-        task.run_wait()
-        self._remove_from_running(task)
-        self._completed_tasks.append(task)
+    def _poll_tasks(self):
+        for part in self._partitions:
+            jobs = []
+            for t in self._scheduler_tasks[part.fullname]:
+                if t.policy_stage == 'compiling':
+                    jobs.append(t.check.build_job)
+                elif t.policy_stage == 'running':
+                    jobs.append(t.check.job)
+
+            part.scheduler.poll(*jobs)
+
+        jobs = []
+        for t in self._local_scheduler_tasks:
+            if t.policy_stage == 'compiling':
+                jobs.append(t.check.build_job)
+            elif t.policy_stage == 'running':
+                jobs.append(t.check.job)
+
+        self.local_scheduler.poll(*jobs)
+
+    def advance_all(self, tasks, timeout=None):
+        t_init = time.time()
+        num_prog = 0
+
+        # progress might remove the tasks that retire or fail
+        for t in list(tasks):
```

Comment:
Just to make sure, you mean to pop them and and add them back in in the end? Tests in general are removed only in case of errors or when they are retired. I will have to change the `_advance_{stage}` methods to return something extra to know whether it should be added back or not.

Reply:
Yes, exactly. But I tried that and it was hanging for some reason. Let's leave it as is for now.